src/WeatherDMD/data_pipeline.py

```diff
@@ -0,0 +1,168 @@
+    lat_slice : slice, optional
+        Slice to extract from the latitude dimension of the dataset.
+    lon_slice : slice, optional
+        Slice to extract from the longitude dimension of the dataset.
+    downsample : int, optional
+        Factor to downsample the dataset in the lat and lon directions. Default is 1. Must be an integer greater than 0.
+
+    Returns
+    -------
+    data : numpy.ndarray
+        Numpy array with the variable extracted from the dataset, with dimensions (time, lat, lon).
+    attrs : dict
+        Dictionary with the attributes of the variable.
+    coords : dict
+        Coordinates of the variable.
+    dims : tuple
+        Tuple with the dimensions of the variable.
+    """
+
+    try:
+        if level is None:
+            data = ds[variable].isel(level=0)
+        else:
+            data = ds[variable].sel(level=level)
+        if lat_slice:
+            data = data.sel(latitude=lat_slice)
+        if lon_slice:
+            data = data.sel(longitude=lon_slice)
+        data = data.coarsen(
+            latitude=downsample, longitude=downsample, boundary="trim"
+        ).mean()
+        attrs = data.attrs
+        coords = dict(data.coords)
+        dims = data.dims
+        data = data.values
+    except Exception as e:
+        print(f"Error converting dataset to numpy array: {e}")
+        raise
+
+    return data, attrs, coords, dims
+
+
+def array_to_dataarray(
+    data: xr.DataArray, attrs: dict, coords: dict, dims: tuple
+) -> xr.DataArray:
+    """
+    Convert numpy array to xarray DataArray.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        Numpy array to convert.
+    attrs : dict
+        Dictionary with the attributes of the variable.
+    coords : dict
+        Coordinates of the variable.
+    dims : tuple
+        Tuple with the dimensions of the variable.
+
+    Returns
+    -------
+    da : xarray.DataArray
+        DataArray created from the numpy array.
+    """
+
+    try:
+        da = xr.DataArray(data, coords=coords, dims=dims)
+        da.attrs = attrs
+    except Exception as e:
+        print(f"Error converting numpy array to DataArray: {e}")
+        raise
+
+    return da
+
+
+def datarray_to_zarr(
+    da: xr.DataArray,
+    variable_name: str,
+    file_name: str = "era5_dmd_forecast",
+    prepend_time: bool = True,
+):
+    """
+    Convert DataArray to Dataset and save it to a Zarr store.
+
+    Parameters
+    ----------
+    da : xarray.DataArray
+        DataArray to save.
+    variable_name : str
+        Name of the variable in the DataArray.
+    file_name : str
+        Name of the file to save. Will be saved in the data/output directory.
+    prepend_time : bool
+        If True, the start and end times of the DataArray will be prepended to the file name.
+    """
+
+    try:
+        if prepend_time:
+            time = da.time.values
+            time_start = np.datetime_as_string(time[0], unit="D")
```

Comment:
In my experience, when messing around with xarray files, one often messes up the "time" dimensions such that its no longer sorted. 

I'd suggest not relying on it being sorted, so the first element might not be the earliest time element and the last one might not be the latest. 

Maybe add a condition that checks that the time dimensions is in fact sorted?